Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +186,161 @@ type family PUnLabel (a :: PLabeledType) :: PType where
  This is how most data structures are stored on-chain.
 -}
 type PDataSum :: [[PLabeledType]] -> PType
-data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+data PDataSum (defs :: [[PLabeledType]]) (s :: S) where 
+  MkPDataSum :: PDataSum' n 'True defs s -> PDataSum defs s 
+
+{- Overview of how this works: 
+
+1) We need constructors for PDataSum in order to define a PlutusType instance. 
+
+2) Ideally, we want the PDataSum data type to be structured so as to enforce the invariant that each sum contains 
+   *exactly one* of its choices
+
+3) We can't do that with PDataSum alone. To enforce those invariants at the type level, we need an index type which encodes 
+   the fact that the sum contains exactly one choice. 
+
+4) PDataSum' is the helper type which allows us to express those invariants. The structure of the type has to be inductive, but 
+   we need four constructors (two Zero / two Suc) to cover each of the possible cases. The Boolean type argument indicates whether or not the 
+   sum contains an element, while the Natural type argument represents the position of the element in the list of choices. 
+
+      a) 'SumZE' represents a one element sum that *does not* contain its element.  
+
+      b) 'SumZF' represents a one element sum that *does* contain its element.
+
+      c) 'InjE' injects an element into a sum which does not contain any elements such that the sum will contain the injected element. 
+
+      d) 'InjF' injects an element into a sum which may or may not contain any elements, and preserves the containment status of 
+         the sum it was injected into. 
+
+      Suggestions for better constructor names are welcome. 
+
+5) PDataSum is a wrapper around a PDataSum' that contains exactly one of its elements. The structure of PDataSum' ensures that any 'PDataSum n 'True' 
+   will contain exactly one of its elements. An empty sum cannot be represented, and there is no way to inject a second element into a sum which already 
+   contains one element such that the resulting sum contains more than one of its elements. 
+
+6) We use a universally quantified continuation to indicate containment. I *think* we have to do it this way. 
+
+7) We need some helper utilities to reconstruct the Sum (which we have to do) in pmatch'. Those utilities are: 
+    
+    a) 'IndexedSum'. This is basically a singleton list a la SList from sop-core, except it can only represent nonempty lists, has an 
+       "explicit" inductive structure, and indicates the position of elements explicitly. 
+
+    b) A 'Length' type family. This doesn't work like 'length' from Prelude: 'Length '[a]' == 0, not 1. 
+
+    c) A type class 'IndexableSum' which is like SListI from sop-core, except for our 'IndexedSum' type. As far as I can tell it is impossible 
+       to reconstruct the sum without an inductive GADT that represents the structure of the sum. We need to conjure that structure out of thin air,
+       so we need a type class. 
```

Comment:
I'm curious why we can't just reuse `NS` and the like from sop-core. Where is the non-empty property being required?

Reply:
NS enforces the nonempty property too, doesn't it? I guess that could work as a replacement for IndexedSum. 

I don't think it works as a replacement for the PDataSum' stuff though. I'm basing that on intuition + failing to get it to work (it was the first thing I tried for PDataSum).